firedrake/mesh.py

```diff
@@ -2036,13 +2036,16 @@ def spatial_index(self):
         coords_min = self._order_data_by_cell_index(column_list, coords_min.dat.data_ro_with_halos)
         coords_max = self._order_data_by_cell_index(column_list, coords_max.dat.data_ro_with_halos)
 
-        # Push max and min out so we can find points on the boundary within
-        # tolerance. Note that if tolerance is too small it might not actually
-        # change the value!
-        if hasattr(self, "tolerance") and self.tolerance is not None:
-            coords_diff = coords_max - coords_min
-            coords_min -= self.tolerance*coords_diff
-            coords_max += self.tolerance*coords_diff
+        # Push max and min out to a hypercube centered at its average so we can
+        # find points on the mesh within tolerance.
+
+        # Note that getattr doesn't work here due to the inheritance games that
+        # are going on in getattr.
+        tolerance = self.tolerance if hasattr(self, "tolerance") else 0.0
+        coords_mid = (coords_max + coords_min)/2
+        d = np.max(coords_max - coords_min, axis=1)[:, None]
+        coords_min = coords_mid - (tolerance + 0.5)*d
+        coords_max = coords_mid + (tolerance + 0.5)*d
```

Comment:
Even if the tolerance is `0.0` this still changes the values of `coords_min` and `coords_max` - I take it that the tolerance is still doing the same job it was before (i.e. pushing `coords_max` and `coords_min` out so we can find points on the boundary within tolerance). To make this clear I think you should break this into two code blocks:
```suggestion
        # find points on the mesh, even for immersed manifolds.
        coords_mid = (coords_max + coords_min)/2
        d = np.max(coords_max - coords_min, axis=1)[:, None]
        coords_min = coords_mid - d/2
        coords_max = coords_mid + d/2
        
        # Push max and min out so we can find points on the boundary within
        # tolerance. Note that getattr doesn't work here due to the 
        # inheritance games that are going on in getattr.
        tolerance = self.tolerance if hasattr(self, "tolerance") else 0.0
        coords_min -= tolerance*d
        coords_max += tolerance*d
```

Reply:
Remember these are whole array operations. We do assume that they are relatively cheap, but I don't think we should be doubling the number just because we assume the reader doesn't know how to expand brackets.

Reply:
Alright, it can be stated in the comments.